### PR TITLE
Complete first assignment

### PR DESCRIPTION
**Pull request by veronikakolejak**

### CODE REVIEW

File: Assignments/Q1Permutation/CheckIfPermutation.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package Q1Permutation;
+
+public class CheckIfPermutation {
+
+    public static boolean arePermutations(String s1, String s2) {
+
+        /* Assuming that there exists a permutation of an empty string */
```
Comment (simon-frankau) [inline Assignments/Q1Permutation/CheckIfPermutation.java]: An empty string being a permutation of itself sounds like a good assumption to me.

File: Assignments/Q1Permutation/CheckIfPermutation.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        if (s1.length() != s2.length()) return false;
+
+        int[] counts = new int[256];
```
Comment (simon-frankau) [inline Assignments/Q1Permutation/CheckIfPermutation.java]: As Java characters are unicode, they're not all in the range 0-255. With appropriate input, this could cause an index-out-of-bounds.

While a hashmap would be slower, it would have the same O-notation complexity and would also handle this case.

File: Assignments/Q1Permutation/CheckIfPermutation.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+        for (char letter : s1.toLowerCase().toCharArray()) {
+            counts[letter]++;
+        }
+
+        for (char letter : s1.toLowerCase().toCharArray()) {
+            counts[letter]--;
+            /* If some counts are higher than 0 then some have to less than 0, thus can check for one of the options only */
```
Comment (simon-frankau) [inline Assignments/Q1Permutation/CheckIfPermutation.java]: Useful comment. This is the tricky part of your algorithm, and documenting the "why" of your code like this is great for a reviewer. Cheers.

File: Assignments/Q1Permutation/CheckIfPermutationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+class CheckIfPermutationTest extends GroovyTestCase {
+    void testArePermutations() {
+        assertTrue(CheckIfPermutation.arePermutations("Silent", "Listen"));
+        assertTrue(CheckIfPermutation.arePermutations("", ""));
+        assertTrue(CheckIfPermutation.arePermutations("siLEnt", "liSteN"));
+        assertFalse(CheckIfPermutation.arePermutations("hello", "helo"));
+        assertFalse(CheckIfPermutation.arePermutations("", "listen"));
+    }
+}
```
Comment (simon-frankau) [inline Assignments/Q1Permutation/CheckIfPermutationTest.groovy]: Nice test cases. I like the coverage of the case hello/helo case, where one's a subset of the other - that's a potential corner case. You might want to test helo/hello too, since s1 and s2 are not treated symmetrically.

File: Assignments/Q2KthToLastElement/KthToLastElement.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package Q2KthToLastElement;
+
+public class KthToLastElement {
+
+    public static Node kthElement(Node head, int k) {
+        if (k < 0) return null;         // Elements with a negative number of elements from the end are undefined
```
Comment (simon-frankau) [inline Assignments/Q2KthToLastElement/KthToLastElement.java]: I'd prefer an exception, which is pretty usual for an index-out-of-bounds kind of case.

File: Assignments/Q2KthToLastElement/KthToLastElement.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+        /* Keep track of two pointers */
+        Node pointer1 = head;
+        Node pointer2 = head;
+
+        /* Move pointer1 exactly k elements ahead */
+        for (int i = 0; i < k; i++) {
+            if (pointer1.next == null) return null;    // Check if k is not larger than the number of elements in the list - 1
```
Comment (simon-frankau) [inline Assignments/Q2KthToLastElement/KthToLastElement.java]: Ditto for here - exception if you're out of bounds.

File: Assignments/Q2KthToLastElement/KthToLastElement.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package Q2KthToLastElement;
+
+public class KthToLastElement {
+
+    public static Node kthElement(Node head, int k) {
+        if (k < 0) return null;         // Elements with a negative number of elements from the end are undefined
+
+        /* Keep track of two pointers */
```
Comment (simon-frankau) [inline Assignments/Q2KthToLastElement/KthToLastElement.java]: Your comments are generally pretty good. This one is a bit obvious.

File: Assignments/Q2KthToLastElement/KthToLastElement.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        for (int i = 1; i < nodeData.length; i++) {
+            Node newNode = new Node(nodeData[i]);
+            node.next = newNode;
+            node = node.next;
+        }
+
+        Node result = kthElement(head, k);
+        if (result != null) System.out.println(result.data);
+    }
+}
```
Comment (simon-frankau) [inline Assignments/Q2KthToLastElement/KthToLastElement.java]: Aw. I'd liked to have seen some more tests in Q2 like the rather good ones you did for Q1.

File: README.md [hunk 9]
```diff
@@ -1 +1,16 @@
...
+Used integer arrays to count occurrences of individual lowercase characters in each string
+
+### Question2
+Method takes a reference to the head node of a linked list and returns the kth to the last element of the linked list.
+Used two pointers of which one was ahead exactly k nodes from the other
+
+  
+*TODO:*
+- Use the Optional class mentioned by Andrew
+- Write unit tests for kthElement method
```
Comment (simon-frankau) [inline README.md]: Ah, yes, that'd be nice!

File: Assignments/Q2KthToLastElement/Node.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package Q2KthToLastElement;
+
+public class Node {
```
Comment (simon-frankau) [inline Assignments/Q2KthToLastElement/Node.java]: It's a bit pedantic, but I think using generics for the Node would have been nice.

**Review by simon-frankau — APPROVED** [review]
Looks like nice, readable, efficient code to me.

I would appreciate it if you use the Google Java code style (braces always used on if blocks, etc.) - it's widely used, and makes my reviewing easier if everyone uses the same style.
